src/transformer_traits.jl

```diff
@@ -0,0 +1,421 @@
+end
+
+#####
+##### Tuples
+#####
+
+function internal_type_structure(::Type{T}, ::StructTypes.ArrayType) where {T<:Tuple}
+    if isconcretetype(T)
+        fields = T <: StructTypes.OrderedStruct ? fieldnames(T) : sorted_field_names(T)
+        return fields, map(field -> fieldtype(T, field), fields)
+    else
+        return nothing
+    end
+end
+
+function internal_type_structure(::Type{T}, ::StructTypes.ArrayType) where {T<:NTuple}
+    return eltype(T)
+end
+
+function stable_hash_helper(x::Tuple, hash_state, context, ::StructTypes.ArrayType)
+    nested_hash_state = start_nested_hash!(hash_state)
+    nested_hash_state = hash_fields(x, fieldnames(typeof(x)), nested_hash_state, context)
+    hash_state = end_nested_hash!(hash_state, nested_hash_state)
+    return hash_state
+end
+
+#####
+##### DictType
+#####
+
+is_ordered(x::AbstractDict) = false
+
+function internal_type_structure(::Type{T}, ::StructTypes.DictType) where {T}
+    return keytype(T), valtype(T)
+end
+
+function transformer(::Type{<:Pair}, ::HashVersion{4})
+    return Transformer(((a, b),) -> (a, b); hoist_type=true)
+end
+
+function stable_hash_helper(x, hash_state, context, ::StructTypes.DictType)
+    pairs = StructTypes.keyvaluepairs(x)
+    nested_hash_state = start_nested_hash!(hash_state)
+
+    pairs = if is_ordered(x)
+        StructTypes.keyvaluepairs(x)
+    else
+        sort!(collect(StructTypes.keyvaluepairs(x)); by=hash_sort_by ∘ first)
+    end
+    transform = transformer(eltype(x), context)::Transformer
+    hash_elements(pairs, nested_hash_state, context, transform)
+
+    hash_state = end_nested_hash!(hash_state, nested_hash_state)
+    return hash_state
+end
+
+#####
+##### CustomStruct
+#####
+
+# we need to hash the type for every instance when we have a CustomStruct; `lowered` could
+# be anything
+function stable_hash_helper(x, hash_state, context, ::StructTypes.CustomStruct)
+    return hash_type_and_value(StructTypes.lower(x), hash_state, context)
+end
+
+#####
+##### Basic data types
+#####
+
+transform_type(::Type{Symbol}) = "Base.Symbol"
+function transformer(::Type{<:Symbol}, ::HashVersion{4})
+    return Transformer(String; hoist_type=true)
+end
+
+function stable_hash_helper(str, hash_state, context, ::StructTypes.StringType)
+    nested_hash_state = start_nested_hash!(hash_state)
+    update_hash!(nested_hash_state, str isa AbstractString ? str : string(str), context)
+    return end_nested_hash!(hash_state, nested_hash_state)
+end
+
+function stable_hash_helper(number::T, hash_state, context,
+                            ::StructTypes.NumberType) where {T}
+    U = StructTypes.numbertype(T)
+    return update_hash!(hash_state, U(number), context)
+end
+
+function stable_hash_helper(bool, hash_state, context, ::StructTypes.BoolType)
+    return update_hash!(hash_state, Bool(bool), context)
+end
+
+# null types are encoded purely by their type hash
+transform_type(::Type{Missing}) = "Base.Missing"
```

Comment:
do we test that `stable_hash(missing; version=4) != stable_hash("Base.Missing"; version=4)`?

Reply:
nope, easy to add (add it should pass, since we'll prepend the string with its type).